benchmarks/CMakeLists.txt

```diff
@@ -0,0 +1,29 @@
+set(SUBSYS_NAME benchmarks)
+set(SUBSYS_DESC "Point cloud library benchmarks")
+set(SUBSYS_DEPS common features search kdtree io)
+set(DEFAULT OFF)
+set(build TRUE)
+set(REASON "Disabled by default")
+PCL_SUBSYS_OPTION(build "${SUBSYS_NAME}" "${SUBSYS_DESC}" ${DEFAULT} "${REASON}")
+PCL_SUBSYS_DEPEND(build "${SUBSYS_NAME}" DEPS ${SUBSYS_DEPS})
+if(NOT build)
+  return()
+endif()
+
+find_package(benchmark REQUIRED)
+add_custom_target(run_benchmarks)
+
+macro(PCL_ADD_BENCHMARK _name)
+  set(options)
+  set(oneValueArgs)
+  set(multiValueArgs ARGUMENTS LINK_WITH)
+  cmake_parse_arguments(PCL_ADD_BENCHMARK "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+  add_executable(benchmark_${_name} ${_name}.cpp)
+  target_link_libraries(benchmark_${_name} benchmark::benchmark ${PCL_ADD_BENCHMARK_LINK_WITH})
+  set_target_properties(benchmark_${_name} PROPERTIES RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
+  add_custom_target(run_benchmark_${_name} benchmark_${_name} ${PCL_ADD_BENCHMARK_ARGUMENTS})
+  add_dependencies(run_benchmarks run_benchmark_${_name})
+endmacro()
```

Comment:
Some notes to the CMake code when I take only a short look at it:
- Why it is a `macro` and not a `function`?
- Source Group is missing (we should create a new source group for it)
- is there a error handling necessary in case the developer is missing a parameter?
- I dislike `add_executable(benchmark_${_name} ${_name}.cpp)` as therefore you have for each benchmark a fixed source name. In general I would prefer to have a benchmark per PCL module (e.g. `pcl_benchmark_common`). In case someone want to run just one test, he can use `--benchmark_filter=...`

Reply:
Thank you for your notes!
- what advantages would a function have over a macro? I mainly oriented myself at the existing macros, e.g. `PCL_ADD_TEST`
- error handling: optional IMO. Then it would make sense to implement it for all macros in pcl_targets.cmake
- my current plan was to have a cpp file for each module (e.g. features.cpp, common.cpp, filters.cpp, ...), and corresponding executables (benchmark_features, benchmark_common, benchmark_filters, ...). If a module ever has that many benchmarks that it makes sense to split it, we would have to rethink the structure
Please also take another look at the current macro after I applied @larshg's suggestions.

Reply:
> * what advantages would a function have over a macro? I mainly oriented myself at the existing macros, e.g. `PCL_ADD_TEST`

It is like in C++: Content of a macro will be injected where it is called, during function are creating a new scope. So in general: Use `function` except there is a reason, that you need really a `macro`

> * error handling: optional IMO. Then it would make sense to implement it for all macros in pcl_targets.cmake

Indeed it makes sense to introduce it for all. Question here was more: Is everything finde, when `LINK_WITH` & `ARGUMENTS` is empty

> * my current plan was to have a cpp file for each module (e.g. features.cpp, common.cpp, filters.cpp, ...), and corresponding executables (benchmark_features, benchmark_common, benchmark_filters, ...). If a module ever has that many benchmarks that it makes sense to split it, we would have to rethink the structure

In general I would prefer to start in a clean way. So your current benchmark would be within `benchmarks/features/normal_3d.cpp`. Reason: Normally when someone is adding sth. he just add it and don't think: Oh well, no the file is to big, let's restructure the project. So it is better to start clean (and it is better for the Git history ;-) ).

Reply:
Please take another look, if I missed any of your suggestions :smiley: